File: src/main/java/com/hjh_database/dz/gui/RecipeEditorGui.java
package com.hjh_database.dz.gui;

import com.hjh_database.Hjh_database;
import com.hjh_database.dz.data.DzRecipe;
import com.hjh_database.util.ItemUtil;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.HandlerList;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.InventoryCloseEvent;
import org.bukkit.event.player.AsyncPlayerChatEvent;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.InventoryHolder;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class RecipeEditorGui implements InventoryHolder, Listener {
    private final Hjh_database plugin;
    private final Player player;
    private final String category;
    private final String editId;
    private final Inventory inv;

    private boolean awaitingChat = false;
    private ItemStack cachedResult;
    private List<ItemStack> cachedIngredients;
    private String cachedId;

    private final int[] INPUT_SLOTS = {11, 12, 13, 14, 15};
    private final int RESULT_SLOT = 24;
    private final int SAVE_BUTTON_SLOT = 49;

    public RecipeEditorGui(Hjh_database plugin, Player player, String category, String editId) {
        this.plugin = plugin;
        this.player = player;
        this.category = category;
        this.editId = editId;

        String title = (editId == null) ? "新建配方" : "编辑: " + editId;
        this.inv = Bukkit.createInventory(this, 54, title);

        setupGui();
        plugin.getServer().getPluginManager().registerEvents(this, plugin);
    }

    private void setupGui() {
        ItemStack bg = new ItemStack(Material.BLACK_STAINED_GLASS_PANE);
        ItemMeta bgm = bg.getItemMeta();
        bgm.setDisplayName("§7");
        bg.setItemMeta(bgm);
        for(int i=0; i<54; i++) {
            if(!isInputOrResult(i)) inv.setItem(i, bg);
        }

        ItemStack save = new ItemStack(Material.LIME_WOOL);
        ItemMeta sm = save.getItemMeta();
        sm.setDisplayName("§a§l[保存并设置参数]");
        sm.setLore(Arrays.asList(
                "§71. 放入成品和材料",
                "§72. 点击此按钮",
                "§73. 在聊天栏输入等级职业等参数"
        ));
        save.setItemMeta(sm);
        inv.setItem(SAVE_BUTTON_SLOT, save);

        if (editId != null) {
            DzRecipe r = plugin.getRecipeManager().getRecipe(category, editId);
            if (r != null) {
                inv.setItem(RESULT_SLOT, r.getResult());
                List<ItemStack> ings = r.getIngredients();
                for(int i=0; i<ings.size() && i<INPUT_SLOTS.length; i++) {
                    inv.setItem(INPUT_SLOTS[i], ings.get(i));
                }
            }
        }
    }

    private boolean isInputOrResult(int slot) {
        if (slot == RESULT_SLOT) return true;
        for(int s : INPUT_SLOTS) if(s == slot) return true;
        return false;
    }

    public void open() { player.openInventory(inv); }
    @Override public Inventory getInventory() { return inv; }

    @EventHandler
    public void onClose(InventoryCloseEvent event) {
        if(event.getInventory().equals(inv)) {
            if (!awaitingChat) {
                HandlerList.unregisterAll(this);
            }
        }
    }

    @EventHandler
    public void onClick(InventoryClickEvent event) {
        if(!event.getInventory().equals(inv)) return;
        int slot = event.getRawSlot();
        if(!isInputOrResult(slot) && slot < 54) event.setCancelled(true);

        if(slot == SAVE_BUTTON_SLOT) {
            event.setCancelled(true);
            initiateSaveProcess();
        }
    }

    private void initiateSaveProcess() {
        ItemStack result = inv.getItem(RESULT_SLOT);
        if (result == null || result.getType() == Material.AIR) {
            player.sendMessage("§c错误：结果槽位为空！");
            return;
        }

        // 核心：识别ID
        String id = (editId != null) ? editId : ItemUtil.getPublicId(result);

        // 如果识别出来是 AIR，说明这既不是 RPG物品，也不是原版物品，这几乎不可能发生，除非 ItemUtil 逻辑有误
        if (id.equals("AIR")) {
            player.sendMessage("§c错误：无法识别结果物品的ID。");
            return;
        }

        List<ItemStack> ings = new ArrayList<>();
        for(int s : INPUT_SLOTS) {
            ItemStack it = inv.getItem(s);
            ings.add(it != null ? it : new ItemStack(Material.AIR));
        }

        this.cachedResult = result;
        this.cachedIngredients = ings;
        this.cachedId = id;
        this.awaitingChat = true;

        player.closeInventory();

        player.sendMessage("§a========================================");
        player.sendMessage("§a正在保存配方，ID识别为: §e" + id);
        if (id.equals(result.getType().name())) {
            player.sendMessage("§7(提示: 这是一个原版物品配方)");
        } else {
            player.sendMessage("§7(提示: 这是一个自定义RPG物品配方)");
        }
        player.sendMessage("§a请在聊天栏输入 4 个参数 (空格分隔):");
        player.sendMessage("§7格式: <职业> <等级> <经验奖励> <执照要求>");
        player.sendMessage("§7示例: 1 10 50 0  (代表:弓手 10级 50经验 0执照)");
        player.sendMessage("§e输入 'cancel' 取消保存");
        player.sendMessage("§a========================================");
    }

    @EventHandler
    public void onChat(AsyncPlayerChatEvent event) {
        if (!event.getPlayer().equals(player) || !awaitingChat) return;

        event.setCancelled(true);
        String msg = event.getMessage().trim();

        if (msg.equalsIgnoreCase("cancel")) {
            player.sendMessage("§c已取消保存。");
            awaitingChat = false;
            HandlerList.unregisterAll(this);
            return;
        }

        String[] args = msg.split("\\s+");
        try {
            int job = args.length > 0 ? Integer.parseInt(args[0]) : -1;
            int lv  = args.length > 1 ? Integer.parseInt(args[1]) : 1;
            int exp = args.length > 2 ? Integer.parseInt(args[2]) : 10;
            int lic = args.length > 3 ? Integer.parseInt(args[3]) : 0;

            DzRecipe recipe = new DzRecipe(
                    cachedId, category, cachedResult, cachedIngredients,
                    job, lv, lic, exp
            );
            plugin.getRecipeManager().saveRecipe(recipe);

            player.sendMessage("§a✔ 配方保存成功！");

            awaitingChat = false;
            HandlerList.unregisterAll(this);

            Bukkit.getScheduler().runTask(plugin, () ->
                    new AdminRecipeListGui(plugin, player, category).open()
            );

        } catch (NumberFormatException e) {
            player.sendMessage("§c格式错误！请输入数字。例如: -1 1 10 0");
        }
    }
}

File: src/main/kotlin/com/hjh_database/dz/gui/RecipeEditorGui.kt
package com.hjh_database.dz.gui

import com.hjh_database.Hjh_database
import com.hjh_database.dz.data.DzRecipe
import com.hjh_database.util.DzUtil
import com.hjh_database.util.ItemUtil
import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.HandlerList
import org.bukkit.event.Listener
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.event.inventory.InventoryCloseEvent
import org.bukkit.event.player.AsyncPlayerChatEvent
import org.bukkit.inventory.Inventory
import org.bukkit.inventory.InventoryHolder
import org.bukkit.inventory.ItemStack
import java.util.ArrayList
import java.util.Arrays

class RecipeEditorGui(
    private val plugin: Hjh_database,
    private val player: Player,
    private val category: String,
    private val editId: String? // 允许为 null，代表新建
) : InventoryHolder, Listener {

    private val inv: Inventory
    private var awaitingChat = false

    // 缓存变量，初始化为 null
    private var cachedResult: ItemStack? = null
    private var cachedIngredients: List<ItemStack>? = null
    private var cachedId: String? = null

    // 缓存原配方对象（编辑时用于 y/Y 快捷保存）
    private var originalRecipe: DzRecipe? = null

    private val INPUT_SLOTS = intArrayOf(11, 12, 13, 14, 15)
    private val RESULT_SLOT = 24
    private val SAVE_BUTTON_SLOT = 49

    init {
        val title = if (editId == null) "新建配方" else "编辑: $editId"
        this.inv = Bukkit.createInventory(this, 54, title)

        setupGui()
        plugin.server.pluginManager.registerEvents(this, plugin)
    }

    private fun setupGui() {
        val bg = ItemStack(Material.BLACK_STAINED_GLASS_PANE)
        val bgm = bg.itemMeta
        if (bgm != null) {
            bgm.setDisplayName("§7")
            bg.itemMeta = bgm
        }
        for (i in 0 until 54) {
            if (!isInputOrResult(i)) inv.setItem(i, bg)
        }

        val save = ItemStack(Material.LIME_WOOL)
        val sm = save.itemMeta
        if (sm != null) {
            sm.setDisplayName("§a§l[保存并设置参数]")
            sm.lore = Arrays.asList(
                "§71. 放入成品和材料",
                "§72. 点击此按钮",
                "§73. 在聊天栏输入锻造等级、职业等参数"
            )
            save.itemMeta = sm
        }
        inv.setItem(SAVE_BUTTON_SLOT, save)

        if (editId != null) {
            val r = plugin.recipeManager.getRecipe(category, editId)
            if (r != null) {
                originalRecipe = r
                inv.setItem(RESULT_SLOT, r.result)
                val ings = r.ingredients
                var i = 0
                while (i < ings.size && i < INPUT_SLOTS.size) {
                    inv.setItem(INPUT_SLOTS[i], ings[i])
                    i++
                }
            }
        }
    }

    private fun isInputOrResult(slot: Int): Boolean {
        if (slot == RESULT_SLOT) return true
        for (s in INPUT_SLOTS) if (s == slot) return true
        return false
    }

    fun open() {
        player.openInventory(inv)
    }

    override fun getInventory(): Inventory {
        return inv
    }

    @EventHandler
    fun onClose(event: InventoryCloseEvent) {
        if (event.inventory == inv) {
            if (!awaitingChat) {
                HandlerList.unregisterAll(this)
            }
        }
    }

    @EventHandler
    fun onClick(event: InventoryClickEvent) {
        if (event.inventory != inv) return
        val slot = event.rawSlot
        if (!isInputOrResult(slot) && slot < 54) event.isCancelled = true

        if (slot == SAVE_BUTTON_SLOT) {
            event.isCancelled = true
            initiateSaveProcess()
        }
    }

    private fun initiateSaveProcess() {
        val result = inv.getItem(RESULT_SLOT)
        if (result == null || result.type == Material.AIR) {
            player.sendMessage("§c错误：结果槽位为空！")
            return
        }

        // 核心：识别ID
        val id = editId ?: ItemUtil.getPublicId(result)

        // 如果识别出来是 AIR，说明这既不是 RPG物品，也不是原版物品，这几乎不可能发生，除非 ItemUtil 逻辑有误
        if (id == "AIR") {
            player.sendMessage("§c错误：无法识别结果物品的ID。")
            return
        }

        val ings: MutableList<ItemStack> = ArrayList()
        for (s in INPUT_SLOTS) {
            val it = inv.getItem(s)
            ings.add(it ?: ItemStack(Material.AIR))
        }

        this.cachedResult = result
        this.cachedIngredients = ings
        this.cachedId = id
        this.awaitingChat = true

        player.closeInventory()

        player.sendMessage("§a========================================")
        player.sendMessage("§a正在保存配方，ID识别为: §e$id")
        if (id == result.type.name) {
            player.sendMessage("§7(提示: 这是一个原版物品配方)")
        } else {
            player.sendMessage("§7(提示: 这是一个自定义RPG物品配方)")
        }

        // 显示当前参数（编辑时显示原有值，新建时显示默认值）
        val prev = originalRecipe
        if (prev != null) {
            player.sendMessage("§6当前配方参数:")
            player.sendMessage("§7  职业=${prev.reqJob}(${DzUtil.getJobName(prev.reqJob)})  锻造等级=${prev.reqForgeLevel}  经验=${prev.expReward}  资质=${prev.reqLicense}")
        } else {
            player.sendMessage("§6默认参数: §7职业=-1(全职业)  锻造等级=1  经验=10  资质=0")
        }

        player.sendMessage("§a请在聊天栏输入 4 个参数 (空格分隔):")
        player.sendMessage("§7格式: <职业> <锻造等级> <经验奖励> <锻造资质要求>")
        player.sendMessage("§7示例: 1 10 50 0  (代表:弓手 锻造等级10 50经验 锻造资质0)")
        player.sendMessage("§e输入 'y' 使用上述参数直接保存")
        player.sendMessage("§e输入 'cancel' 取消保存")
        player.sendMessage("§a========================================")
    }

    @EventHandler
    fun onChat(event: AsyncPlayerChatEvent) {
        if (event.player != player || !awaitingChat) return

        event.isCancelled = true
        val msg = event.message.trim()

        if (msg.equals("cancel", ignoreCase = true)) {
            player.sendMessage("§c已取消保存。")
            awaitingChat = false
            HandlerList.unregisterAll(this)
            return
        }

        // y/Y 快捷保存: 使用之前配置好的数据
        if (msg.equals("y", ignoreCase = true)) {
            val prev = originalRecipe
            val job: Int
            val lv: Int
            val exp: Int
            val lic: Int

            if (prev != null) {
                // 编辑已有配方：复用原配方参数
                job = prev.reqJob
                lv = prev.reqForgeLevel
                exp = prev.expReward
                lic = prev.reqLicense
            } else {
                // 新建配方：使用默认值
                job = -1
                lv = 1
                exp = 10
                lic = 0
            }

            saveRecipe(job, lv, exp, lic)
            return
        }

        val args = msg.split("\\s+".toRegex()).toTypedArray()
        try {
            val job = if (args.isNotEmpty()) args[0].toInt() else -1
            val lv = if (args.size > 1) args[1].toInt() else 1
            val exp = if (args.size > 2) args[2].toInt() else 10
            val lic = if (args.size > 3) args[3].toInt() else 0

            saveRecipe(job, lv, exp, lic)

        } catch (e: NumberFormatException) {
            player.sendMessage("§c格式错误！请输入数字。例如: -1 1 10 0")
        }
    }

    /**
     * 执行配方保存的公共逻辑
     */
    private fun saveRecipe(job: Int, lv: Int, exp: Int, lic: Int) {
        if (cachedId != null && cachedResult != null && cachedIngredients != null) {
            val recipe = DzRecipe(
                cachedId!!, category, cachedResult!!, cachedIngredients!!,
                job, lv, lic, exp
            )
            plugin.recipeManager.saveRecipe(recipe)
            player.sendMessage("§a✔ 配方保存成功！")
            player.sendMessage("§7参数: 职业=${job}(${DzUtil.getJobName(job)}) 锻造等级=$lv 经验=$exp 资质=$lic")
        } else {
            player.sendMessage("§c错误：缓存数据丢失，保存失败。")
        }

        awaitingChat = false
        HandlerList.unregisterAll(this)

        // 回到主线程打开 GUI
        Bukkit.getScheduler().runTask(plugin, Runnable {
            AdminRecipeListGui(plugin, player, category).open()
        })
    }
}
